### DISCUSSION TITLE
Bug report - model fails at random places if using CUDA device

### DISCUSSION THREAD

**Discussion by S-trace** [snakers4/silero-models#225] [Q&A] [answered]
## 🐛 Bug

When I'm trying to use torch.device('cuda', 0) on model load - it fails at random places of text.
It does not fails if I run the same script with torch.device('cpu')

## To Reproduce

1. Start TTS with  torch.device('cuda', 0)
2. Wait for some time

The error message I'm getting each time is:
```
Traceback (most recent call last):
  File "<torch_package_0>.multi_acc_v3_package.py", line 338, in apply_tts
  File "/usr/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File ".data/ts_code/code/__torch__/tacotron2/forward_layers.py", line 18, in forward
      _5 = torch.select(x, 0, i)
      _6 = torch.add(torch.select(dur, 0, i), 0.5)
      x_exp = torch.repeat_interleave(_5, torch.to(_6, 4), 0)
              ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
      _7 = torch.append(x_expanded, x_exp)
    return _0(x_expanded, True, 0., )

Traceback of TorchScript, original code (most recent call last):
  File "../tacotron2/forward_layers.py", line 32, in forward
        x_expanded = []
        for i in range(x.size(0)):
            x_exp = torch.repeat_interleave(x[i], (dur[i] + 0.5).long(), dim=0)
                    ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
            x_expanded.append(x_exp)
        x_expanded = pad_sequence(x_expanded, padding_value=0., batch_first=True)
RuntimeError: repeats can not be negative


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/s-trace/development/tts/./tts.py", line 387, in <module>
    main()
  File "/home/s-trace/development/tts/./tts.py", line 73, in main
    process_tts(tts_model, preprocessed_lines, input_filename, wave_file_size_limit, preprocessed_text_len)
  File "/home/s-trace/development/tts/./tts.py", line 371, in process_tts
    audio = tts_model.apply_tts(text=line,
  File "<torch_package_0>.multi_acc_v3_package.py", line 340, in apply_tts
Exception: Model couldn't generate your text, probably it's too long
```
The complete script: https://github.com/S-trace/silero_tts_standalone/blob/master/tts.py

## Expected behavior

TTS should not fail with CUDA.

## Environment

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.26.3
Libc version: glibc-2.37

Python version: 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201] (64-bit runtime)
Python platform: Linux-6.2.10-1-MANJARO-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 12.1.66
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050 Ti
Nvidia driver version: 530.41.03
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.8.0
/usr/lib/libcudnn_adv_infer.so.8.8.0
/usr/lib/libcudnn_adv_train.so.8.8.0
/usr/lib/libcudnn_cnn_infer.so.8.8.0
/usr/lib/libcudnn_cnn_train.so.8.8.0
/usr/lib/libcudnn_ops_infer.so.8.8.0
/usr/lib/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          12
On-line CPU(s) list:             0-11
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
CPU family:                      6
Model:                           158
Thread(s) per core:              2
Core(s) per socket:              6
Socket(s):                       1
Stepping:                        10
CPU(s) scaling MHz:              76%
CPU max MHz:                     4100,0000
CPU min MHz:                     800,0000
BogoMIPS:                        4401,32
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       192 KiB (6 instances)
L1i cache:                       192 KiB (6 instances)
L2 cache:                        1,5 MiB (6 instances)
L3 cache:                        9 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-11
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[conda] Could not collect

## Additional context
I'm using "xenia" voice.

**Comment by snakers4** [comment]
>  it fails at random places of text.
> Exception: Model couldn't generate your text, probably it's too long

Does this happen if you limit input text length to 512, 768 or 1024 chars?

**Comment by S-trace** [comment]
> Does this happen if you limit input text length to 512, 768 or 1024 chars?

It happen with limit of 1024 chars or 941 chars, but not with 512 and 768 chars.

With 1024 chars limit i've got a wonderful tracebacks pack:
```
Traceback (most recent call last):
  File "<torch_package_0>.multi_acc_v3_package.py", line 338, in apply_tts
  File "/usr/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File ".data/ts_code/code/__torch__/silero_vocoder/jit_model/___torch_mangle_180.py", line 56, in forward
    _8 = torch.to(durs_rate, torch.device(device))
    _9 = torch.to(pitch_coefs, torch.device(device))
    _10 = (tts_model).forward(_6, _7, sr, symb_durs, _8, _9, gt_durs, gt_pitch, device, )
           ~~~~~~~~~~~~~~~~~~ <--- HERE
    audio, audio_lengths, = _10
    return (audio, audio_lengths)
  File ".data/ts_code/code/__torch__/silero_vocoder/jit_model/___torch_mangle_179.py", line 153, in forward
      pitch_hat = unchecked_cast(Tensor, gt_pitch)
    tacotron = self.tacotron
    mel_outputs = (tacotron).forward(sequence, speaker_ids, orig_mask, dur_hat, pitch_hat, )
                   ~~~~~~~~~~~~~~~~~ <--- HERE
    if torch.__isnot__(symb_durs4, None):
      symb_durs20 = unchecked_cast(Dict[int, int], symb_durs4)
  File ".data/ts_code/code/__torch__/jit_forward_model/___torch_mangle_178.py", line 43, in forward
    encoder_outputs_expanded = (len_reg).forward(cond_encoder_outputs0, dur_hat, )
    decoder = self.decoder
    outputs_expanded = (decoder).forward(encoder_outputs_expanded, None, )
                        ~~~~~~~~~~~~~~~~ <--- HERE
    lin = self.lin
    mel_outputs = (lin).forward(outputs_expanded, )
  File ".data/ts_code/code/__torch__/tacotron2/fastpitch_layers.py", line 118, in forward
    x8 = torch.transpose(x, 0, 1)
    pos_encoder = self.pos_encoder
    x9 = (pos_encoder).forward(x8, )
          ~~~~~~~~~~~~~~~~~~~~ <--- HERE
    x10 = torch.transpose(x9, 0, 1)
    pre_vanilla_layers = self.pre_vanilla_layers
  File ".data/ts_code/code/__torch__/tacotron2/fastpitch_layers.py", line 42, in forward
    _0 = torch.slice(pe, 0, None, torch.size(x, 0))
    _1 = torch.mul(scale, torch.slice(_0, 1))
    x7 = torch.add(x, _1)
         ~~~~~~~~~ <--- HERE
    dropout = self.dropout
    return (dropout).forward(x7, )

Traceback of TorchScript, original code (most recent call last):
  File "../../silero_vocoder/jit_model.py", line 198, in forward
        sequence, symb_durs, durs_rate, pitch_coefs = self.merge_batch_model(acc_sentences, break_lens, prosody_rates, prosody_pitches)
    
        audio, audio_lengths = self.tts_model(sequence=sequence.to(device),
                               ~~~~~~~~~~~~~~ <--- HERE
                                              speaker_ids=speaker_ids.to(device),
                                              sr=sr,
  File "../../silero_vocoder/jit_model.py", line 457, in forward
            pitch_hat = gt_pitch
    
        mel_outputs = self.tacotron(sequence, speaker_ids, orig_mask, dur_hat, pitch_hat)
                      ~~~~~~~~~~~~~ <--- HERE
        if symb_durs is not None and len(symb_durs) > 0:
            mel_outputs = self.fx_pauses(mel_outputs, dur_hat, symb_durs)
  File "/home/keras/notebook/nvme2/islanna/silero_vocoder/tacotron2/jit_forward_model.py", line 110, in forward
    
        # [B, Lexp, Denc]
        outputs_expanded = self.decoder(encoder_outputs_expanded,
                           ~~~~~~~~~~~~ <--- HERE
                                        src_pad_mask=None)
    
  File "../tacotron2/fastpitch_layers.py", line 335, in forward
        x = x.transpose(0, 1)
        # [L, B, d_m]
        x = self.pos_encoder(x)
            ~~~~~~~~~~~~~~~~ <--- HERE
        # [B, L, d_m]
        x = x.transpose(0, 1)
  File "../tacotron2/fastpitch_layers.py", line 53, in forward
    def forward(self, x: torch.Tensor) -> torch.Tensor:         # shape: [T, N]
        x = x + self.scale * self.pe[:x.size(0), :]
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
        return self.dropout(x)
RuntimeError: The size of tensor a (5020) must match the size of tensor b (5000) at non-singleton dimension 0


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/s-trace/development/tts/./tts.py", line 392, in <module>
    main()
  File "/home/s-trace/development/tts/./tts.py", line 74, in main
    process_tts(tts_model, preprocessed_lines, input_filename, wave_file_size_limit, preprocessed_text_len)
  File "/home/s-trace/development/tts/./tts.py", line 376, in process_tts
    audio = tts_model.apply_tts(text=line,
  File "<torch_package_0>.multi_acc_v3_package.py", line 340, in apply_tts
Exception: Model couldn't generate your text, probably it's too long
```

**Comment by snakers4** [comment, marked as answer]
You should just limit the input length

![image](https://user-images.githubusercontent.com/12515440/232253483-91a62d24-9db9-427d-920a-4b1d505b4d35.png)
